### PR TITLE
Refactor infra of Go 1.18 fuzzers

### PR DESCRIPTION
**Pull request by AdamKorcz**
`go-118-fuzz-build` is undergoing refactoring to fix some runtime issues affecting Rekor and other projects. Currently none of Rekors fuzzers run. This PR has been tested on all of Rekors fuzzers that will run with this PR.

This PR modifies the infra OSS-Fuzz build script to adapt to the upstream changes.

Cloning the `dev` branch of `go-118-fuzz-build` is a temporary solution until it gets merged into `main` upstream.
Cloning go-118-fuzz-build in each build script is also a temporary measure until `dev` gets merged into `main`.

Signed-off-by: AdamKorcz <adam@adalogics.com>

### CODE REVIEW

**Comment by AdamKorcz** [comment]
@Navidem This one should go green.

**Review by oliverchang — COMMENTED** [review]
thank you for fixing! would you mind summarizing the changes and what this fixes in the PR description?

**Comment by AdamKorcz** [comment]
> thank you for fixing! would you mind summarizing the actual fixes in the PR description to give us a bit more context?
> 
> e.g. I see `rewrite_go_fuzz_harness` got removed. What upstream change led to removing this?

Sure, the changes in this PR are 3-fold:

1. Update the compile_native_go_fuzzer script to accomodate upstream changes.
2. Update projects with new changes.
3. Temporarily disable the coverage build of native go fuzzers.

**1: Update the `compile_native_go_fuzzer` script to accomodate upstream changes.**
Several features that were found in `compile_native_go_fuzzer` have been moved to the `go-118-fuzz-build` binary. The two biggest changes are:
1. **Harness modification:** Some OSS-Fuzz projects have moved their test utils out of the fuzzers, and the rewrite-script in `compile_native_go_fuzzer` does not modify those test utils. Istio is an example of that which results in crashes like this: https://oss-fuzz.com/testcase-detail/6582039348641792. `go-118-fuzz-build` has been updated to modify the fuzzer and all of its dependencies by traversing the fuzzers and its dependecies' AST. [This file](https://github.com/AdamKorcz/go-118-fuzz-build/blob/dev/file_walker.go) in `go-118-fuzz-build` does that. The `addimport` binary has also been removed.
2. **Better support for testing APIs**. More OSS-Fuzz projects are now writing fuzzers with std lib testing utils like `t.Skip()` which were not supported in `go-118-fuzz-build`. I have added support for these in `go-118-fuzz-build` in a way that makes sense for libFuzzer. This can be found here: https://github.com/AdamKorcz/go-118-fuzz-build/tree/dev/testing. This is for example an issue for Rekor which breaks the fuzzers at runtime.

**2: Update projects with new changes**
This part adjusts the projects in OSS-Fuzz using `github.com/AdamKorcz/go-118-fuzz-build` in their build scripts. The biggest change is to the paths:

1. `.../testingtypes` -> `.../testing`
2. `.../utils` -> `.../testing`

**3: Temporarily disable the coverage build of native go fuzzers**
To land this, I have disabled the coverage build for native 1.18 fuzzers. The coverage build should ideally also be reduced; Std lib Go fuzzers don't need a wrapper and can be run directly as unit tests. There is currently work being done in `go-118-fuzz-build` to add a tool to rewrite corpus files from libFuzzer format to native Go fuzzing format which will allow users to take a testcase of a crash in OSS-Fuzz and test it in their native Go fuzzers. This will also allow us to run the fuzzers as unit tests without the coverage wrapper. Once this is ready, I will update the OSS-Fuzz coverage build which should result in removing [this file](https://github.com/google/oss-fuzz/blob/master/infra/base-images/base-builder-go/native_ossfuzz_coverage_runner.go).

**Comment by AdamKorcz** [comment]
@oliverchang As discussed, I have fixed the coverage build instead of deferring it to a future PR. Could you review?

File: infra/base-images/base-runner/coverage [hunk 1]
```diff
@@ -279,7 +291,10 @@ done
 wait
 
 if [[ $FUZZING_LANGUAGE == "go" ]]; then
+  ls $DUMPS_DIR
```
Comment (oliverchang) [inline infra/base-images/base-runner/coverage:294]: Is this leftover debugging logging?

File: infra/base-images/base-builder/install_go.sh [hunk 2]
```diff
@@ -28,11 +28,7 @@ go install github.com/mdempsky/go114-fuzz-build@latest
 ln -s $GOPATH/bin/go114-fuzz-build $GOPATH/bin/go-fuzz
 
 cd /tmp
-git clone https://github.com/AdamKorcz/go-118-fuzz-build
+git clone --branch=dev https://github.com/AdamKorcz/go-118-fuzz-build
```
Comment (oliverchang) [inline infra/base-images/base-builder/install_go.sh:31]: Should we expect the dev branch changes to be merged into the main branch soon?

**Review by oliverchang — APPROVED** [review]
thanks! This LGTM with just two more comments/questions.

File: infra/base-images/base-runner/coverage (hunk 1 above)
Comment (AdamKorcz) [reply infra/base-images/base-runner/coverage]: It was. Thank you for catching that.

File: infra/base-images/base-builder/install_go.sh (hunk 2 above)
Comment (AdamKorcz) [reply infra/base-images/base-builder/install_go.sh:31]: Yes. That should happen by end of month.

**Review by oliverchang — APPROVED** [review]

**Comment by oliverchang** [comment]
/gcbrun trial_build.py smt helm moby go-dns go-coredns istio --sanitizer address --fuzzing-engine libfuzzer

**Comment by oliverchang** [comment]
From the trial build, 

istio is failing with many instances of:

```
+ compile_native_go_fuzzer istio.io/istio/pilot/pkg/security/authz/builder FuzzBuildTCP FuzzBuildTCP
pkg/config/mesh/fuzz_test.go_fuzz.go:3:8: no required module provides package github.com/AdamKorcz/go-118-fuzz-build/testing; to add it:
	go get github.com/AdamKorcz/go-118-fuzz-build/testing
```

helm is failing with many instances of:

```
Running go-fuzz -tags gofuzz -func FuzzIndexDirectory -o fuzz_index_directory.a helm.sh/helm/v3/pkg/repo
+ compile_go_fuzzer helm.sh/helm/v3/pkg/repo FuzzDownloadIndexFile fuzz_download_index_file
/usr/local/bin/compile_go_fuzzer: line 28: cd: /root/go/src/helm.sh/helm/v3/pkg/repo: No such file or directory
```

These appear to be passing normally without this PR. 

@AdamKorcz is this expected? can you please investigate?

File: projects/istio/build.sh [hunk 3]
```diff
@@ -15,6 +15,15 @@
 #
 ################################################################################
 
+cd $SRC
+git clone --branch=dev https://github.com/AdamKorcz/go-118-fuzz-build $SRC/go-118-fuzz-build
+cd $SRC/istio
+go mod edit -replace github.com/AdamKorcz/go-118-fuzz-build=/src/go-118-fuzz-build
```
Comment (oliverchang) [inline projects/istio/build.sh:21]: Can you add TODOs here in these project build.sh scripts to remove once upstreamed?

**Comment by oliverchang** [comment]
> > From the trial build,
> > istio is failing with many instances of:
> > ```
> > + compile_native_go_fuzzer istio.io/istio/pilot/pkg/security/authz/builder FuzzBuildTCP FuzzBuildTCP
> > pkg/config/mesh/fuzz_test.go_fuzz.go:3:8: no required module provides package github.com/AdamKorcz/go-118-fuzz-build/testing; to add it:
> > 	go get github.com/AdamKorcz/go-118-fuzz-build/testing
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > helm is failing with many instances of:
> > ```
> > Running go-fuzz -tags gofuzz -func FuzzIndexDirectory -o fuzz_index_directory.a helm.sh/helm/v3/pkg/repo
> > + compile_go_fuzzer helm.sh/helm/v3/pkg/repo FuzzDownloadIndexFile fuzz_download_index_file
> > /usr/local/bin/compile_go_fuzzer: line 28: cd: /root/go/src/helm.sh/helm/v3/pkg/repo: No such file or directory
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > These appear to be passing normally without this PR.
> > @AdamKorcz is this expected? can you please investigate?
> 
> Yes it is expected.
> 
> Istio and Helm need similar updates as the OSS-Fuzz build scripts(change `/utils` to `/testing`). I have pushed an example of that directly in the Istio build and pointed Helm to a branch where it is fixed. Istio needs a bit more work with [these 3 fuzzers](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/fuzz_test.go), but I think it is better to do that upstream. The current state of this PR should show that Istios and Helms errors quoted above have been fixed; In the logs, I only see that Istio is failing:
> 
> **libfuzzer, address, x86_64** `Failed projects: istio`
> 
> **libfuzzer, coverage, x86_64** `Failed projects: istio`
> 
> Naturally I will adjust the upstream build scripts of all projects running native Go fuzzers.

Thanks. If other breakages are expected, can we please have those fixes ready to go with this PR? IT wouldn't be good if we start spamming oss-fuzz users with build failure emails because of our own breakage.

File: projects/istio/build.sh (hunk 3 above)
Comment (AdamKorcz) [reply projects/istio/build.sh]: Created this instead: https://github.com/istio/istio/pull/42028

**Comment by AdamKorcz** [comment]
I have gone through all Go projects and found that 3 (besides those fixed in this PR) would break from this PR. I have created a draft with a fix on those projects:

- Containerd: https://github.com/containerd/containerd/pull/7677
- Istio: https://github.com/istio/istio/pull/42028
- Cosmos-SDK: https://github.com/cosmos/cosmos-sdk/pull/13889

I have removed the Istio fix from this PR.

**Comment by oliverchang** [comment]
OK, thanks! Please merge this whenever you're ready. We should aim to merge the three fixes you listed as soon as this one lands.

**Comment by RiSKeD** [comment]
@AdamKorcz Would it be possible for you to take a look at the build script of the u-root project as well? 
I tried updating the build script after the changes, but it fails to build some of the targets.

**Comment by AdamKorcz** [comment]
> @AdamKorcz Would it be possible for you to take a look at the build script of the u-root project as well? I tried updating the build script after the changes, but it fails to build some of the targets.

Will do

**Comment by catenacyber** [comment]
https://github.com/google/oss-fuzz/issues/9808 looks caused by this

Why should we have coverage only for native fuzzing @AdamKorcz ?

**Pull request merged**